processor/metricstransformprocessor/metrics_transform_processor_helper.go

```diff
@@ -0,0 +1,375 @@
+		for _, p := range points {
+			if aggrType == Sum || aggrType == Average {
+				intVal += p.GetInt64Value()
+			} else if aggrType == Max {
+				if p.GetInt64Value() <= intVal {
+					continue
+				}
+				intVal = p.GetInt64Value()
+			}
+
+		}
+		if aggrType == Average {
+			intVal /= int64(len(points))
+		}
+		return &metricspb.Point_Int64Value{Int64Value: intVal}, nil, nil
+	case metricspb.MetricDescriptor_GAUGE_DOUBLE, metricspb.MetricDescriptor_CUMULATIVE_DOUBLE:
+		for _, p := range points {
+			if aggrType == Sum || aggrType == Average {
+				doubleVal += p.GetDoubleValue()
+			} else if aggrType == Max {
+				if p.GetDoubleValue() <= doubleVal {
+					continue
+				}
+				doubleVal = p.GetDoubleValue()
+			}
+
+		}
+		if aggrType == Average {
+			doubleVal /= float64(len(points))
+		}
+		return nil, &metricspb.Point_DoubleValue{DoubleValue: doubleVal}, nil
+	case metricspb.MetricDescriptor_GAUGE_DISTRIBUTION, metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		if aggrType != Sum {
+			break
+		}
+		boundsToPoints := mtp.groupPointsByBounds(points)
+		for _, v := range boundsToPoints {
+			var distVal *metricspb.DistributionValue
+			for _, p := range v {
+				if distVal == nil {
+					distVal = p.GetDistributionValue()
+					continue
+				}
+				distVal = mtp.computeDistVals(distVal, p.GetDistributionValue())
+			}
+			distVals = append(distVals, &metricspb.Point_DistributionValue{DistributionValue: distVal})
+		}
+		return nil, nil, distVals
+
+	}
+	return nil, nil, nil
+}
+
+// groupPointsByBounds groups distribution value points by the bounds to further determine the aggregatability because distribution value points can only be aggregated if they have match bounds
+// Returns a map with groups of aggregatable distribution value points
+func (mtp *metricsTransformProcessor) groupPointsByBounds(points []*metricspb.Point) map[string][]*metricspb.Point {
+	boundsToPoints := make(map[string][]*metricspb.Point)
+	for _, p := range points {
+		boundsKey := mtp.boundsToString(p.GetDistributionValue().BucketOptions.GetExplicit().Bounds)
+		if groupedPoints, ok := boundsToPoints[boundsKey]; ok {
+			boundsToPoints[boundsKey] = append(groupedPoints, p)
+		} else {
+			boundsToPoints[boundsKey] = []*metricspb.Point{p}
+		}
+	}
+	return boundsToPoints
+}
+
+// boundsToString converts bounds slice into a string
+// Returns a string that represents this bounds
+func (mtp *metricsTransformProcessor) boundsToString(bounds []float64) string {
+	var boundsKey string
+	for _, b := range bounds {
+		boundsKey += fmt.Sprintf("%f-", b)
+	}
+	return boundsKey
+}
+
+// sumOfSquaredDeviation calculates the combined SumOfSquaredDeviation from the ones in val1 and val2
+// Formula derived and extended from https://math.stackexchange.com/questions/2971315/how-do-i-combine-standard-deviations-of-two-groups
+// SSDcomb = SSD1 + n(ave(x) - ave(z))^2 + SSD2 +n(ave(y) - ave(z))^2
```

Comment:
My maths is not as good as yours but I'm really struggling to think about how to feel confident in the correctness of this function. I can't even figure out how you derived your function from the linked one 😄 

How confident are you that this is correct? It may be worth writing some test cases to cover this calculation explicitly.

Reply:
I am pretty sure but not 100% sure about the math either haha, but I am pretty confident about my derivation steps. I currently have a test case with one being 3, and another one being 4. I set this up purposefully for easy calculation because in that test case the difference of the averages is exactly 0, which eliminates 2 of the terms in there. I can also come up with a more concrete testcase with specific points in each set for testing purposes. What do you think?

Reply:
This seems like something that would be good to test with [property based tests](https://medium.com/criteo-labs/introduction-to-property-based-testing-f5236229d237), but don't worry about that for now.

I do think it's worth creating a specific test case for this function though.

Reply:
Added the example case.